File: traffic_cypher_in_C/src_c/frame_sampler.c
#include "frame_sampler.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

struct frame_capture {
    FILE   *fp;
    pid_t   pid;
    int     pipe_fd;
    uint64_t sequence;
};

frame_capture_t *frame_capture_start(const char *stream_url) {
    fprintf(stderr, "[INFO] Starting ffmpeg frame capture at 1 FPS\n");

    int pipefd[2];
    if (pipe(pipefd) < 0) {
        perror("pipe");
        return NULL;
    }

    pid_t pid = fork();
    if (pid < 0) {
        perror("fork");
        close(pipefd[0]);
        close(pipefd[1]);
        return NULL;
    }

    if (pid == 0) {
        /* Child: redirect stdout to pipe */
        close(pipefd[0]);
        dup2(pipefd[1], STDOUT_FILENO);
        close(pipefd[1]);

        /* Redirect stderr to /dev/null for cleaner output */
        FILE *devnull = fopen("/dev/null", "w");
        if (devnull) {
            dup2(fileno(devnull), STDERR_FILENO);
            fclose(devnull);
        }

        execlp("ffmpeg", "ffmpeg",
            "-reconnect", "1",
            "-reconnect_streamed", "1",
            "-reconnect_delay_max", "5",
            "-i", stream_url,
            "-vf", "fps=1,scale=320:240",
            "-f", "image2pipe",
            "-vcodec", "ppm",
            "-an",
            "-loglevel", "error",
            "pipe:1",
            (char *)NULL);

        perror("execlp ffmpeg");
        _exit(127);
    }

    /* Parent */
    close(pipefd[1]);

    frame_capture_t *cap = (frame_capture_t *)calloc(1, sizeof(frame_capture_t));
    cap->pipe_fd = pipefd[0];
    cap->fp = fdopen(pipefd[0], "r");
    cap->pid = pid;
    cap->sequence = 0;

    if (!cap->fp) {
        perror("fdopen");
        close(pipefd[0]);
        free(cap);
        return NULL;
    }

    return cap;
}

/* Read exactly n bytes from file stream */
static int read_exact(FILE *fp, uint8_t *buf, size_t n) {
    size_t total = 0;
    while (total < n) {
        size_t r = fread(buf + total, 1, n - total, fp);
        if (r == 0) return -1; /* EOF or error */
        total += r;
    }
    return 0;
}

int frame_capture_read(frame_capture_t *cap, frame_t *frame) {
    if (!cap || !cap->fp) return -1;

    /* Read PPM header: "P6\n<width> <height>\n<maxval>\n" */
    char header_line[256];

    /* Line 1: "P6" */
    if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    /* Skip comment lines */
    while (header_line[0] == '#') {
        if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    }
    if (strncmp(header_line, "P6", 2) != 0) return -1;

    /* Line 2: "<width> <height>" (may have comments before) */
    if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    while (header_line[0] == '#') {
        if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    }
    uint32_t w = 0, h = 0;
    if (sscanf(header_line, "%u %u", &w, &h) != 2) return -1;

    /* Line 3: maxval */
    if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    while (header_line[0] == '#') {
        if (!fgets(header_line, sizeof(header_line), cap->fp)) return -1;
    }

    /* Read pixel data */
    size_t pixel_bytes = (size_t)w * h * 3;
    uint8_t *data = (uint8_t *)malloc(pixel_bytes);
    if (!data) return -1;

    if (read_exact(cap->fp, data, pixel_bytes) < 0) {
        free(data);
        return -1;
    }

    cap->sequence++;
    frame->width = w;
    frame->height = h;
    frame->data = data;
    frame->data_len = pixel_bytes;
    frame->sequence = cap->sequence;

    return 0;
}

void frame_capture_stop(frame_capture_t *cap) {
    if (!cap) return;
    if (cap->fp) {
        fclose(cap->fp);
        cap->fp = NULL;
    }
    if (cap->pid > 0) {
        kill(cap->pid, SIGTERM);
    }
    free(cap);
}

pid_t frame_capture_pid(const frame_capture_t *cap) {
    return cap ? cap->pid : 0;
}

File: traffic_cypher_in_Rust/src/frame_sampler.rs
use anyhow::{Context, Result, bail};
use tokio::io::AsyncReadExt;
use tokio::process::Command;
use tokio::sync::mpsc;
use tracing::{debug, error, info};

/// A single captured frame with its raw pixel data.
#[derive(Clone)]
pub struct Frame {
    pub width: u32,
    pub height: u32,
    pub data: Vec<u8>, // raw RGB pixel bytes
    pub sequence: u64,
}

/// Start ffmpeg as a subprocess reading from the resolved stream URL,
/// outputting 1 PPM frame per second to stdout.
/// Sends parsed frames through the returned channel.
pub async fn start_frame_capture(
    stream_url: &str,
    sender: mpsc::Sender<Frame>,
) -> Result<tokio::process::Child> {
    info!("Starting ffmpeg frame capture at 1 FPS");

    let mut child = Command::new("ffmpeg")
        .args([
            "-reconnect", "1",
            "-reconnect_streamed", "1",
            "-reconnect_delay_max", "5",
            "-i", stream_url,
            "-vf", "fps=1,scale=320:240",
            "-f", "image2pipe",
            "-vcodec", "ppm",
            "-an",           // no audio
            "-loglevel", "error",
            "pipe:1",
        ])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .context("Failed to spawn ffmpeg. Is it installed?")?;

    let stdout = child.stdout.take()
        .context("Failed to capture ffmpeg stdout")?;

    // Spawn a task to continuously read and parse PPM frames
    tokio::spawn(async move {
        if let Err(e) = read_ppm_stream(stdout, sender).await {
            error!("Frame reader error: {}", e);
        }
    });

    Ok(child)
}

/// Read PPM (P6) frames from a byte stream.
/// PPM format: "P6\n<width> <height>\n<maxval>\n<RGB bytes>"
async fn read_ppm_stream(
    mut reader: impl AsyncReadExt + Unpin,
    sender: mpsc::Sender<Frame>,
) -> Result<()> {
    let mut sequence: u64 = 0;
    let mut buf = Vec::with_capacity(320 * 240 * 3 + 256);

    loop {
        // Read PPM header byte by byte
        let (width, height, header_len) = match read_ppm_header(&mut reader, &mut buf).await {
            Ok(v) => v,
            Err(e) => {
                debug!("PPM header read ended: {}", e);
                break;
            }
        };

        let pixel_bytes = (width * height * 3) as usize;
        let mut pixel_data = vec![0u8; pixel_bytes];
        reader.read_exact(&mut pixel_data).await
            .context("Failed to read PPM pixel data")?;

        sequence += 1;

        let frame = Frame {
            width,
            height,
            data: pixel_data,
            sequence,
        };

        debug!("Captured frame #{} ({}x{})", sequence, width, height);

        if sender.send(frame).await.is_err() {
            info!("Frame receiver dropped, stopping capture");
            break;
        }

        buf.clear();
    }

    Ok(())
}

/// Parse a PPM P6 header from the stream.
/// Returns (width, height, header_byte_count).
async fn read_ppm_header(
    reader: &mut (impl AsyncReadExt + Unpin),
    buf: &mut Vec<u8>,
) -> Result<(u32, u32, usize)> {
    buf.clear();

    // Read until we have at least the header (P6\n<w> <h>\n<maxval>\n)
    // We read byte-by-byte to avoid over-reading into pixel data
    let mut newline_count = 0;
    let mut byte = [0u8; 1];

    loop {
        let n = reader.read(&mut byte).await?;
        if n == 0 {
            bail!("EOF while reading PPM header");
        }
        buf.push(byte[0]);

        // Skip comment lines
        if byte[0] == b'\n' {
            // Check if next line is a comment
            newline_count += 1;
            if newline_count >= 3 {
                break;
            }
        }
    }

    let header = String::from_utf8_lossy(buf);
    let mut lines: Vec<&str> = header.lines().collect();

    // Filter out comment lines (starting with #)
    lines.retain(|l| !l.starts_with('#'));

    if lines.len() < 3 || lines[0].trim() != "P6" {
        bail!("Invalid PPM header: {:?}", header);
    }

    let dims: Vec<&str> = lines[1].trim().split_whitespace().collect();
    if dims.len() < 2 {
        bail!("Invalid PPM dimensions: {}", lines[1]);
    }

    let width: u32 = dims[0].parse().context("Invalid width")?;
    let height: u32 = dims[1].parse().context("Invalid height")?;

    Ok((width, height, buf.len()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_parse_ppm_header() {
        // Construct a minimal PPM P6 frame
        let header = b"P6\n4 2\n255\n";
        let pixels = vec![128u8; 4 * 2 * 3]; // 4x2 RGB
        let mut data = Vec::new();
        data.extend_from_slice(header);
        data.extend_from_slice(&pixels);

        let mut cursor = &data[..];
        let mut buf = Vec::new();
        let (w, h, _) = read_ppm_header(&mut cursor, &mut buf).await.unwrap();
        assert_eq!(w, 4);
        assert_eq!(h, 2);
    }
}
